test/jdk/java/net/httpclient/lib/jdk/httpclient/test/lib/http2/Http2TestServerConnection.java

```diff
@@ -729,6 +729,12 @@ headers, rspheadersBuilder, uri, bis, getSSLSession(),
                     }
                 }
                 throw closed;
+            } finally {
+                // isEof() for if END_STREAM has been seen.
+                // q.size() > 0 for if frames in Queue have not been read.
+                if (bis instanceof BodyInputStream inputStream && (!inputStream.isEof() || inputStream.q.size() > 0)) {
```

Comment:
Caught an error with this code here when `bis` is assigned to `NullInputStream.INSTANCE` above. Need to put in an additional check for that.

Reply:
To add on to the explanation given by the in-line comments. 

- If an END_STREAM is not seen by the `BodyInputStream`, `bis`, then the client has stopped transmitting for some reason. For example if the sendWindow has been exceeded and the client is waiting for WINDOW_UPDATE frames. This can occur if the InputStream is not fully consumed by the exchange handler too.
- It is possible too that the EOF has been seen but that the `BodyInputStream's` Queue is non empty. This is why an || is used instead of an &&.

Through testing this seems to be the most stable solution, this code usually should only get triggered in test cases including a test handler which does not consume the input stream.

Reply:
This is strange and probably too late: when you reach here the bos has probably already been closed by the handler, so setting the flag will have no effect in the usual case.

Reply:
In the usual case yes, if the handler reads the stream this code will have no effect. I think thats what we want though as we don't want the sending of reset frame to be triggered if the server handler reads/closes the stream. It should only be sent when the stream has not been closed/consumed by the handler and the exchanges send window has been exceeded.

Reply:
I see. That sounds right. Can you had a comment above line 736 to explain that?

Reply:
Sounds good.